### DISCUSSION TITLE
Deploy Next.js (app router) + API to A2 Hosting - cPanel Hosting (Node.js application)

### DISCUSSION THREAD

**Discussion by aliawanai** [vercel/next.js#52826] [Help] [answered]
### Summary

Hi, 

### i need help as i am stuck on how to deploy next js app to cPanel hosting which supports Node.js applications.

**Here is what i have**

1. Next.js app which uses next app router.
2. Next js based API routes.
3. Using Prisma with MySQL
4. App is working in localhost

**Here is what i want**
1. I have a A2 Hosting with Cpanel
2. it supports Node.js applications
3. I have MySQL Database
4. i want to deploy my next.js app on this hosting

what is the best approach to do this? without losing + points of Next.js?

**Comment by robertpiosik** [comment]
Next.js in itself won't differ from any other node application. E.g. you can follow any tutorial for express.js app but where it says to run command `node index.js` you do `next start`.

**Reply by aliawanai** [reply]
Node js applications in Cpanel requires a file like app.js to run the node application. i don't have root access to run npm start command. i only have access to cPanel

**Reply by robertpiosik** [reply]
Then I don't think you will be able to use app router, unfortunately.

**Reply by aliawanai** [reply]
should i change hosting plan or make changes in project to use default pages router? what do you suggest? considering scalability?

**Reply by robertpiosik** [reply, marked as answer]
Documentation mentions that switching to custom server disables some important Next.js features like incremental static generation. I would consider typical approach with PM2 process manager as you will be interested in running many processes for greater responsiveness on multicore cpu. About using pages router - I think app router is really fun to use. Good luck!

**Comment by abass-dev** [comment]
In this context, it's preferable to adopt a static development approach. Modify your configuration within the `next.config` file by setting `"output"` to `"export"`. This is the strategy I employ on my website, abassdev.com. If you're utilizing dynamic routing, make sure to export `generateStaticParams()` in your pages. However, a drawback arises when fetching data from an API. In such cases, you'll have to rebuild and deploy whenever your incoming data changes. Alternatively, you can sidestep dynamic routing, opting for a React hook with a client-server setup. This way, you won't need to rebuild and deploy your site every time your database undergoes changes.
